File: research/experiments/sweep/runs/opt-medium-cpp-sonnet45/tests/vec_ops_test.cpp
#include "doctest.h"
#include "vec_ops.h"
#include <cmath>

using namespace optimization;

TEST_CASE("vec-ops: dot product") {
    CHECK(dot({1, 2, 3}, {4, 5, 6}) == doctest::Approx(32));
    CHECK(dot({0, 0}, {1, 1}) == doctest::Approx(0));
}

TEST_CASE("vec-ops: norm") {
    CHECK(norm({3, 4}) == doctest::Approx(5));
    CHECK(norm({0, 0, 0}) == doctest::Approx(0));
}

TEST_CASE("vec-ops: normInf") {
    CHECK(normInf({1, -3, 2}) == doctest::Approx(3));
    CHECK(normInf({0, 0}) == doctest::Approx(0));
}

TEST_CASE("vec-ops: scale") {
    Vector result = scale({1, 2}, 3);
    CHECK(result[0] == doctest::Approx(3));
    CHECK(result[1] == doctest::Approx(6));

    result = scale({1, 2}, 0);
    CHECK(result[0] == doctest::Approx(0));
    CHECK(result[1] == doctest::Approx(0));
}

TEST_CASE("vec-ops: add") {
    Vector result = add({1, 2}, {3, 4});
    CHECK(result[0] == doctest::Approx(4));
    CHECK(result[1] == doctest::Approx(6));
}

TEST_CASE("vec-ops: sub") {
    Vector result = sub({3, 4}, {1, 2});
    CHECK(result[0] == doctest::Approx(2));
    CHECK(result[1] == doctest::Approx(2));
}

TEST_CASE("vec-ops: negate") {
    Vector result = negate({1, -2});
    CHECK(result[0] == doctest::Approx(-1));
    CHECK(result[1] == doctest::Approx(2));
}

TEST_CASE("vec-ops: clone") {
    Vector original = {1, 2};
    Vector cloned = clone(original);
    CHECK(cloned[0] == doctest::Approx(1));
    CHECK(cloned[1] == doctest::Approx(2));

    // Verify it's a distinct array
    cloned[0] = 99;
    CHECK(original[0] == doctest::Approx(1));
}

TEST_CASE("vec-ops: zeros") {
    Vector result = zeros(3);
    CHECK(result.size() == 3);
    CHECK(result[0] == doctest::Approx(0));
    CHECK(result[1] == doctest::Approx(0));
    CHECK(result[2] == doctest::Approx(0));
}

TEST_CASE("vec-ops: addScaled") {
    Vector result = addScaled({1, 2}, {3, 4}, 2);
    CHECK(result[0] == doctest::Approx(7));
    CHECK(result[1] == doctest::Approx(10));
}

TEST_CASE("vec-ops: purity checks") {
    Vector a = {1, 2};
    Vector b = {3, 4};

    // add should not modify inputs
    Vector result = add(a, b);
    CHECK(a[0] == doctest::Approx(1));
    CHECK(b[0] == doctest::Approx(3));

    // scale should not modify input
    Vector v = {1, 2};
    result = scale(v, 3);
    CHECK(v[0] == doctest::Approx(1));
}

File: research/experiments/sweep/runs/opt-medium-go-sonnet45/vec_ops_test.go
package optimization

import (
	"math"
	"testing"
)

func TestDot(t *testing.T) {
	tests := []struct {
		name string
		a, b []float64
		want float64
	}{
		{"basic", []float64{1, 2, 3}, []float64{4, 5, 6}, 32},
		{"zeros", []float64{0, 0}, []float64{1, 1}, 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := Dot(tt.a, tt.b)
			if got != tt.want {
				t.Errorf("Dot(%v, %v) = %v, want %v", tt.a, tt.b, got, tt.want)
			}
		})
	}
}

func TestNorm(t *testing.T) {
	tests := []struct {
		name string
		v    []float64
		want float64
	}{
		{"3-4 triangle", []float64{3, 4}, 5},
		{"zeros", []float64{0, 0, 0}, 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := Norm(tt.v)
			if math.Abs(got-tt.want) > 1e-10 {
				t.Errorf("Norm(%v) = %v, want %v", tt.v, got, tt.want)
			}
		})
	}
}

func TestNormInf(t *testing.T) {
	tests := []struct {
		name string
		v    []float64
		want float64
	}{
		{"mixed signs", []float64{1, -3, 2}, 3},
		{"zeros", []float64{0, 0}, 0},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := NormInf(tt.v)
			if got != tt.want {
				t.Errorf("NormInf(%v) = %v, want %v", tt.v, got, tt.want)
			}
		})
	}
}

func TestScale(t *testing.T) {
	tests := []struct {
		name string
		v    []float64
		s    float64
		want []float64
	}{
		{"basic", []float64{1, 2}, 3, []float64{3, 6}},
		{"zero scalar", []float64{1, 2}, 0, []float64{0, 0}},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := Scale(tt.v, tt.s)
			if !vecEqual(got, tt.want) {
				t.Errorf("Scale(%v, %v) = %v, want %v", tt.v, tt.s, got, tt.want)
			}
			// Check immutability
			if &got[0] == &tt.v[0] {
				t.Errorf("Scale modified input vector")
			}
		})
	}
}

func TestAdd(t *testing.T) {
	a := []float64{1, 2}
	b := []float64{3, 4}
	want := []float64{4, 6}

	got := Add(a, b)
	if !vecEqual(got, want) {
		t.Errorf("Add(%v, %v) = %v, want %v", a, b, got, want)
	}

	// Check immutability
	if a[0] != 1 || b[0] != 3 {
		t.Errorf("Add modified input vectors")
	}
}

func TestSub(t *testing.T) {
	a := []float64{3, 4}
	b := []float64{1, 2}
	want := []float64{2, 2}

	got := Sub(a, b)
	if !vecEqual(got, want) {
		t.Errorf("Sub(%v, %v) = %v, want %v", a, b, got, want)
	}
}

func TestNegate(t *testing.T) {
	v := []float64{1, -2}
	want := []float64{-1, 2}

	got := Negate(v)
	if !vecEqual(got, want) {
		t.Errorf("Negate(%v) = %v, want %v", v, got, want)
	}
}

func TestClone(t *testing.T) {
	original := []float64{1, 2}
	cloned := Clone(original)

	if !vecEqual(cloned, original) {
		t.Errorf("Clone(%v) = %v, want %v", original, cloned, original)
	}

	// Verify it's a distinct array
	cloned[0] = 999
	if original[0] == 999 {
		t.Errorf("Clone did not create a distinct array")
	}
}

func TestZeros(t *testing.T) {
	got := Zeros(3)
	want := []float64{0, 0, 0}

	if !vecEqual(got, want) {
		t.Errorf("Zeros(3) = %v, want %v", got, want)
	}
}

func TestAddScaled(t *testing.T) {
	a := []float64{1, 2}
	b := []float64{3, 4}
	s := 2.0
	want := []float64{7, 10}

	got := AddScaled(a, b, s)
	if !vecEqual(got, want) {
		t.Errorf("AddScaled(%v, %v, %v) = %v, want %v", a, b, s, got, want)
	}
}

// Helper function to compare vectors with tolerance
func vecEqual(a, b []float64) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if math.Abs(a[i]-b[i]) > 1e-10 {
			return false
		}
	}
	return true
}
